File: cpp/알고리즘별/Graph/2178(미로_탐색).cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int N, M;
string S;
int dr[4] = { -1, 1, 0, 0 };
int dc[4] = { 0, 0, -1, 1 };
vector <vector <pair <int, int>>> v;	// first는 입력 값, second는 cnt값
vector <vector <bool>> visited;

void BFS(int row, int col, int cnt);

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> N >> M;

	v.resize(N, vector <pair <int, int>>(M));
	visited.resize(N, vector <bool>(M));

	for (int i = 0; i < N; i++) {
		cin >> S;

		for (int j = 0; j < M; j++) {
			v[i][j].first = S[j] - '0';
			v[i][j].second = 0;
		}
	}

	BFS(0, 0, 0);

	return 0;
}

void BFS(int row, int col, int cnt) {
	queue <pair <pair <int, int>, int>> q;

	v[row][col].second = 1;
	q.push({ {row, col}, cnt + 1 });
	visited[row][col] = true;

	while (!q.empty()) {
		int r = q.front().first.first;
		int c = q.front().first.second;
		int count = q.front().second;

		q.pop();

		for (int i = 0; i < 4; i++) {
			int nr = r + dr[i];
			int nc = c + dc[i];

			if (nr < 0 || nc < 0 || nr >= N || nc >= M || v[nr][nc].first == 0) {
				continue;
			}

			if (!visited[nr][nc]) {
				v[nr][nc].second = count + 1;
				q.push({ { nr, nc }, count + 1 });
				visited[nr][nc] = true;
			}
		}
	}

	cout << v[N - 1][M - 1].second;
}

File: java/알고리즘별/Graph/2178(미로_탐색).java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedWriter bw = new BufferedWriter (new OutputStreamWriter(System.out));
    static BufferedReader br = new BufferedReader (new InputStreamReader(System.in));

    static int N;
    static int M;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    static int [][] arr;
    static int [][] map;
    static boolean [][] visited;

    public static class Triple {
        int first;
        int second;
        int third;

        public Triple(int first, int second, int third) {
            this.first = first;
            this.second = second;
            this.third = third;
        }

        public int getFirst() {
            return first;
        }

        public int getSecond() {
            return second;
        }

        public int getThird() {
            return third;
        }
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        map = new int [N][M];
        arr = new int [N][M];
        visited = new boolean [N][M];

        for (int i = 0; i < N; i++) {
            String s = br.readLine();

            for (int j = 0; j < s.length(); j++) {
                arr[i][j] = s.charAt(j) - '0';
            }
        }

        BFS(0, 0, 0);

        bw.write(Integer.toString(map[N - 1][M - 1]));
        br.close();
        bw.flush();
        bw.close();
    }

    public static void BFS(int row, int col, int cnt) throws IOException {
        map[row][col] = cnt + 1;
        Queue<Triple> q = new LinkedList<>();
        q.add(new Triple(row, col, cnt + 1));
        visited[row][col] = true;

        while(q.isEmpty() == false) {
            int r = q.peek().getFirst();
            int c = q.peek().getSecond();
            int cnt2 = q.peek().getThird();

            q.remove();

            for (int i = 0; i < 4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];

                if (nr < 0 || nc < 0 || nr >= N || nc >= M || visited[nr][nc] == true || arr[nr][nc] == 0) {
                    continue;
                }

                map[nr][nc] = cnt2 + 1;
                q.add(new Triple(nr, nc, cnt2 + 1));
                visited[nr][nc] = true;
            }
        }
    }
}
